modules/apps/29-fee/ibc_module.go

```diff
@@ -171,32 +171,100 @@ func (im IBCModule) OnChanCloseConfirm(
 }
 
 // OnRecvPacket implements the IBCModule interface.
+// If fees are not enabled, this callback will default to the ibc-core packet callback
 func (im IBCModule) OnRecvPacket(
 	ctx sdk.Context,
 	packet channeltypes.Packet,
 	relayer sdk.AccAddress,
-) ibcexported.Acknowledgement {
-	// TODO: Implement fee specific logic if fee is enabled for the given channel
-	return im.app.OnRecvPacket(ctx, packet, relayer)
+) exported.Acknowledgement {
+	if !im.keeper.IsFeeEnabled(ctx, packet.SourcePort, packet.SourceChannel) {
+		return im.app.OnRecvPacket(ctx, packet, relayer)
+	}
+
+	ack := im.app.OnRecvPacket(ctx, packet, relayer)
+
+	forwardRelayer, found := im.keeper.GetCounterpartyAddress(ctx, relayer.String())
+	if !found {
+		forwardRelayer = ""
+	}
+
+	return types.IncentivizedAcknowledgement{
+		Result:                ack.Acknowledgement(),
+		ForwardRelayerAddress: forwardRelayer,
+	}
 }
 
 // OnAcknowledgementPacket implements the IBCModule interface
+// If fees are not enabled, this callback will default to the ibc-core packet callback
 func (im IBCModule) OnAcknowledgementPacket(
 	ctx sdk.Context,
 	packet channeltypes.Packet,
 	acknowledgement []byte,
 	relayer sdk.AccAddress,
 ) error {
-	// TODO: Implement fee specific logic if fee is enabled for the given channel
-	return im.app.OnAcknowledgementPacket(ctx, packet, acknowledgement, relayer)
+	if !im.keeper.IsFeeEnabled(ctx, packet.SourcePort, packet.SourceChannel) {
+		return im.app.OnAcknowledgementPacket(ctx, packet, acknowledgement, relayer)
+	}
+
+	ack := types.IncentivizedAcknowledgement{}
+	if err := types.ModuleCdc.UnmarshalJSON(acknowledgement, &ack); err != nil {
+		return sdkerrors.Wrapf(err, "cannot unmarshal ICS-29 incentivized packet acknowledgement: %v", ack)
+	}
+
+	packetId := channeltypes.NewPacketId(packet.SourceChannel, packet.SourcePort, packet.Sequence)
+	identifiedPacketFee, found := im.keeper.GetFeeInEscrow(ctx, packetId)
+
+	// return underlying callback if no fee found for given packetID
+	if !found {
+		return im.app.OnAcknowledgementPacket(ctx, packet, ack.Result, relayer)
+	}
+	// cache context before trying to distribute the fee
+	cacheCtx, writeFn := ctx.CacheContext()
+
+	err := im.keeper.DistributeFee(cacheCtx, sdk.AccAddress(identifiedPacketFee.RefundAddress), sdk.AccAddress(ack.ForwardRelayerAddress), relayer, packetId)
+	// emit the error in event if any
+	ctx.EventManager().EmitEvents(cacheCtx.EventManager().Events())
```

Comment:
Should these cached events only be emitted if the err is nil and `writeFn()` is being called?
Wouldn't it be misleading to have cached events from bank `SendCoins` emitted if there's an error and `writeFn()` is skipped?

Reply:
I see your point. But I also think it's important to bubble the errors out somewhere, otherwise it'll be v difficult to debug.

Reply:
The events from `DistributeFee` should only be emitted if the err is nil. Otherwise a new event should be emitted with the error message from `DistributeFee`

Events are not typically used for error debugging. The error string should provide most of the information used for debugging

By default, I wouldn't expect the error to be emitted in the event (this is an exceptional situation)